Martinh/wrapped token demo (#237)
wormhole-foundation/wormhole-docs: tutorials/by-product/wormhole-sdk/tokens-via-token-bridge.md

Before transferring a token across chains, you must verify whether it is wrapped on the destination chain. If a token is not native to the target blockchain, Wormhole wraps it into a compatible representation for receiving and using on that chain.

In this section, you'll create a script that checks if a token is wrapped on the destination chain. If it isn't, the script will generate an attestation and complete the wrapping process. This example checks explicitly whether Arbitrum Sepolia has a wrapped version on Base Sepolia and, if not, registers it on the destination chain.

1. Create the `create-wrapped.ts` file in the `src` directory

    ```bash
    mkdir -p src/scripts
    touch src/scripts/create-wrapped.ts
    ```

2. Open `create-wrapped.ts` and import the required modules

    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-3.ts:1:6"
    ```

     - **`evm`** - this import is for working with EVM-compatible chains, like Avalanche, Ethereum, Base Sepolia, and more
     - **`solana`** - this adds support for Solana, a non-EVM chain
     - **`sui`** - support for Sui chain, another non-EVM chain

>      - `helpers.ts` methods defined in the previous section

Reviewer: ```suggestion
     - `helpers.ts` - methods defined in the previous section
```

9. **Submit the attestation on the destination chain** - if the token is not already wrapped, submit the attestation to create the wrapped version on the destination chain

    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-3.ts:66:72"
    ```

     - `submitAttestation(vaa, recipient)` - submits the signed VAA to wrap the token
     - `signSendWait(destChain, subAttestation, signer)` - sends the attestation transaction and waits for confirmation

10. **Wait for the wrapped asset to be available** - after submitting the attestation, poll until the wrapped token is available on the destination chain

    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-3.ts:77:91"
    ```

     If the token is not found, it logs a message and retries after a short delay. Once the wrapped asset is detected, its address is returned.

You can find the complete create-wrapped token script below:

??? code "`create-wrapped.ts`"
    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-3.ts"

>     ```

Reviewer 2: @martin0995 this comment still applies. We should avoid using sentences that introduce elements that are hidden by default. Just remove the introduction sentence

    ```

    !!! note
        You can replace `'Testnet'` with `'Mainnet'` if you want to perform transfers on Mainnet.

4. **Define the source and destination chains** - specify the source chain and destination chain, and retrieve the token ID

    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-2.ts:12:14"
    ```

     - `srcChain` - specifies the source chain, in this case, Arbitrum Sepolia
     - `destChain` - specifies the destination chain, in this case, Base Sepolia
     - `token` - retrieves the native token ID on the source chain

5. **Set up the destination chain signer** - configure the signer for the destination chain and retrieve the Token Bridge instance

    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-2.ts:17:19"
    ```

     - `gasLimit` - defines an optional gas limit for EVM chains
     - `getSigner(destChain, gasLimit)` - retrieves the signer for the destination chain
     - `getTokenBridge()` - retrieves the Token Bridge instance for the destination chain

> 6. **Check if the token is already wrapped on the destination chain** - before creating an attestation, first verify if the token already exists as a wrapped asset

Reviewer 2: ```suggestion
6. **Check if the token is wrapped on the destination chain** - verify if the token already exists as a wrapped asset before creating an attestation
```

    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-2.ts:22:32"
    ```

    If the token is already wrapped, the script exits, and you may proceed to the [next section](/docs/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/#token-transfers). Otherwise, an attestation must be generated.

> 7. **Set up the source chain signer** - configure the signer for the source chain

Reviewer 2: Repetitive. You could answer why they need the signer here "...to create the attestation transaction"

    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-2.ts:35:35"
    ```

> 8. **Create an attestation transaction** - if the token is not already wrapped, generate an attestation to allow it to be registered on the destination chain
> 
>     ```typescript
>     --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-2.ts:38:47"
>     ```
> 
>      - `createAttestation` - generates an attestation for the token on the source chain
>      - `signSendWait` - submits the attestation transaction and waits for confirmation
>      - `txid` - stores the transaction ID, which can be used later to verify the attestation

Reviewer 2: ```suggestion
8. **Create an attestation transaction** - generate and send an attestation for the token on the source chain to register it on the destination chain, then save the transaction ID to verify the attestation in the next step

    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-2.ts:38:47"
    ```
```

Again, this is a little bit repetitive

> 9. **Retrieve the signed VAA** - once the attestation transaction is confirmed, retrieve the signed VAA. This serves as cryptographic proof that the Wormhole network recognizes the attestation
> 
>     ```typescript
>     --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-2.ts:50:59"
>     ```
> 
>      - `parseTransaction(txid)` - extracts Wormhole messages from the attestation transaction
>      - `timeout` - defines how long to wait for the VAA before failing
>      - `getVaa` - retrieves the signed VAA for the attestation

Reviewer 2: ```suggestion
9. **Retrieve the signed VAA** - once the attestation transaction is confirmed, use `parseTransaction(txid)` to extract Wormhole messages, then retrieve the signed VAA from the messages. The timeout defines how long to wait for the VAA before failure

    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-2.ts:50:59"
    ```
```

> 10. **Submit the attestation on the destination chain** - if the token is not already wrapped, submit the attestation to create the wrapped version on the destination chain
> 
>     ```typescript
>     --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-2.ts:66:71"
>     ```
> 
>      - `submitAttestation(vaa, recipient)` - submits the signed VAA to wrap the token
>      - `signSendWait(destChain, subAttestation, signer)` - sends the attestation transaction and waits for confirmation

Reviewer 2: ```suggestion
10. **Submit the attestation on the destination chain** - submit the signed VAA using `submitAttestation(vaa, recipient)` to create the wrapped token on the destination chain, then send the transaction and await confirmation

    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-2.ts:66:71"
    ```
```

> 11. **Wait for the wrapped asset to be available** - after submitting the attestation, poll until the wrapped token is available on the destination chain

Reviewer 2: ```suggestion
11. **Wait for the wrapped asset to be available** - poll until the wrapped token is available on the destination chain
```

    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-2.ts:75:89"
    ```

    If the token is not found, it logs a message and retries after a short delay. Once the wrapped asset is detected, its address is returned.

You can find the complete create-wrapped token script below:

??? code "Complete script"
    ```typescript
    --8<-- "code/tutorials/by-product/wormhole-sdk/tokens-via-token-bridge/token-bridge-2.ts"
    ```

### Run the Wrapped Token Creation

Once the script is ready, execute it with:

```bash
npx tsx src/scripts/create-wrapped.ts
```

> 

Reviewer 2: Now what? Do they need to do anything with the stuff that outputs from the script? Or they're good to go on to the next section?
